contract: Fix challenge logic

sebastianst opened the pull request:
Resolves #57, resolves #65

sebastianst committed:
contract: Implicitly freeze on recovery withdrawals

sebastianst commented:
Forgot to regenerate bindings :joy: Will only remove the `freeze` call though as the remaining API is unchanged.

I am also surprised that the tests work, don't we test freezing somewhere?

sebastianst committed:
contract: Make sure only active system participants can challenge.

The contract had two bugs:

1. Anyone could post a challenge even if they never entered the system.
This would effectively freeze the contract because the operator couldn't
answer a challenge by an unknown user.
2. Once the contract was frozen, a user could withdraw with any balance
proof, not just the frozen epoch's one.

Both bugs are now mitigated by, when posting a challenge, having the
user either post the balance proof of the last sealed epoch, proofing
that they are part of the system; or it is implicitly required that the
user made a deposit in the challenged epoch. This also means that the
to be recovered value is already determined and stored when challenging.

ggwpez commented on contracts/contracts/Erdstall.sol:
```diff
@@ -190,6 +199,10 @@ contract Erdstall {
         return epoch()-3;
     }
 
+    function sealedEpoch() internal view returns (uint64) {
```
I will add a `SealedEpoch(uint64)` to `params.go`, you can then later-on merge it into this commit.

ggwpez replied:
This comment was still in my review buffer, outdated now.

sebastianst committed:
contract: Add missing withdrawFrozen function for non-challengers

matthiasgeihs commented on contracts/contracts/Erdstall.sol:
```diff
@@ -101,63 +99,99 @@ contract Erdstall {
-    // posted.
-    // Can be called by any user if the prior phase has an open challenge.
-    // This must be called in the phase following the challenged exit phase.
-    function freeze() external {
```
What do you think about keeping a method `freeze` for convenience and clarity? It's essential that `withdrawChallenge` is called in the immediately following epoch after the challenged epoch, right? With withdrawals you could usually also wait until later periods, right? With the current naming, somebody could think he can also just withdraw later, but then it's too late.

sebastianst replied:
Already there, now called `ensureFrozen public`. Note that `public` is both, `internal` and `external`.

matthiasgeihs commented on contracts/contracts/Erdstall.sol:
```diff
@@ -101,63 +99,99 @@ contract Erdstall {
+    //
+    // Implicitly calls ensureFrozen to ensure that the contract state is set to
+    // frozen if the last epoch has an unanswered challenge.
+    function withdrawChallenge() external {
```
In the interest of simplicity, what it be possible to remove `withdrawChallenge` and only use `withdrawFrozen`?

sebastianst replied:
It would be and I thought about it, but it would be less optimal because then passing the proof and verifying it would happen twice, which is quite expensive.

matthiasgeihs commented on contracts/contracts/Erdstall.sol:
```diff
@@ -190,6 +224,10 @@ contract Erdstall {
         return epoch()-3;
     }
 
+    function sealedEpoch() internal view returns (uint64) {
```
Seems redundant to `freezingEpoch`?

sebastianst replied:
It's on purpose. Semantically clearer.

matthiasgeihs reviewed (commented):
Had a brief look at the contract. Generally seems ok to me, just a few comments.

ggwpez committed:
:sparkles: [op] Add option to ignore DP and BP generation.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:truck: Make ToEthBals public.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:sparkles: [tee] Add convenience conversion functions.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:art: Use ConfirmTransaction instead of waitMined.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:wrench: Update README, configs + deps.

Delete cmd/operator/config.json in favor or demo/config.json.
Use account-indices 2 and 3, since 0 and 1 are used by OP+enclave.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

ggwpez committed:
:alien: [client,test/user] Implement protocol changes.

Make the FreezeWatcher stoppable.

Signed-off-by: Oliver Tale-Yazdi <oliver@perun.network>

sebastianst commented on tee/tee.go:
```diff
@@ -196,3 +198,11 @@ func (b *Block) GobDecode(blob []byte) error {
 	return nil
 }
+
+func (b *Balance) ToEthBals() bindings.ErdstallBalance {
```
I'd name it `ToEthBal` or `ToEthBalance` since it's singular

sebastianst commented on operator/operator.go:
```diff
@@ -241,12 +241,7 @@ func (operator *Operator) handleChallengedEvent(c challengedEvent) error {
-		Value:   (*big.Int)(balanceProof.Balance.Value),
-	}
-
+	balance := balanceProof.Balance.ToEthBals()
 	tx, err := operator.contract.Exit(tr, balance, balanceProof.Sig)
```
You could move it directly into the argument since `balance` is only moved once.

sebastianst commented on client/client.go:
```diff
@@ -599,7 +633,10 @@ func (c *Client) listenOnChain() error {
-	go c.FrozenWatcher()
+	frozenCtx, cancel := context.WithCancel(c.Ctx())
+	c.stopFrozenWatcher = cancel
+	go c.FrozenWatcher(frozenCtx)
```
Since you start the frozen watcher from this listener, do we really want to make the `FrozenWatcher` method public?

ggwpez replied:
private now

sebastianst commented on client/client.go:
```diff
@@ -439,14 +431,26 @@ func (c *Client) handleFrozen(event *bindings.ErdstallFrozen) {
+
+func (c *Client) withdrawFrozen(status chan *CmdStatus, epoch uint64) {
+	c.balMtx.Lock()
+	defer c.balMtx.Unlock()
```
Do you really need to lock the mutex for the whole duration of the function? And if it's a `RWLock` a `RLock` would be sufficient.

ggwpez replied:
fixed

sebastianst commented on client/client.go:
```diff
@@ -494,26 +498,55 @@ func (c *Client) CmdChallenge(status chan *CmdStatus, args ...string) {
+	// Get the BP of the sealed epoch.
+	lastBlock = atomic.LoadUint64(&c.lastBlock) + 1
+	sealed := c.params.SealedEpoch(lastBlock)
+	c.balMtx.Lock()
```
RLock?

sebastianst commented on client/client.go:
```diff
@@ -494,26 +498,55 @@ func (c *Client) CmdChallenge(status chan *CmdStatus, args ...string) {
 	}
 
+	// Get the BP of the sealed epoch.
+	lastBlock = atomic.LoadUint64(&c.lastBlock) + 1
```
It would be nice to put all `atomic.LoadUint64(&c.lastBlock)` into its own getter `LastBlock`. Maybe also add another getter for the +1 version called `ActiveBlock`.

ggwpez replied:
Added both

sebastianst commented on client/client.go:
```diff
@@ -494,26 +498,55 @@ func (c *Client) CmdChallenge(status chan *CmdStatus, args ...string) {
+	c.waitForReponse(status, tx.BlockNumber.Uint64())
+}
 
+func (c *Client) waitForReponse(status chan *CmdStatus, block uint64) {
```
```suggestion
func (c *Client) waitForResponseOrWithdraw(status chan *CmdStatus, block uint64) {
```
there was a missing `s` in `Response` and since this function is actually also doing something (withdrawing), I'd state this explicitly in its name.

ggwpez replied:
\+ added a comment

sebastianst commented on client/client.go:
```diff
@@ -397,8 +393,8 @@ func (c *Client) BalanceProofWatcher() {
-func (c *Client) FrozenWatcher() {
-	ctx, cancel := context.WithCancel(c.Ctx())
+func (c *Client) FrozenWatcher(_ctx context.Context) {
+	ctx, cancel := context.WithCancel(_ctx)
 	defer cancel()
```
you already pass a cancel-context whose cancel function you set as the watcher canceller. So you don't need this cancel wrapper.

But I think it's actually better to not pass a context to the `FrozenWatcher` at all and set
`ctx, c.stopFrozenWatcher = context.WithCancel(c.Ctx())` here directly.

ggwpez replied:
Yes good idea
